Make string serialization of C++ FunctionSchema consistent with torchgen.model.FunctionSchema

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77926

There is a discrepency between the string representation of C++ FunctionSchema and torchgen.model.FunctionSchema.
The latter will not add parenthesis around the returned types if that a single item,
but the C++ FunctionSchema always add the parenthesis.

Make them consistent so we can convert one type to the other via its string representation and parse method.

Differential Revision: [D36535924](https://our.internmc.facebook.com/intern/diff/D36535924/)

shunting314 committed:
Make string serialization of C++ FunctionSchema consistent with torchgen.model.FunctionSchema

There is a discrepency between the string representation of C++ FunctionSchema and torchgen.model.FunctionSchema.
The latter will not add parenthesis around the returned types if that a single item,
but the C++ FunctionSchema always add the parenthesis.

Make them consistent so we can convert one type to the other via its string representation and parse method.

Differential Revision: [D36535924](https://our.internmc.facebook.com/intern/diff/D36535924/)

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -35,7 +35,10 @@ inline std::ostream& operator<<(std::ostream& out, const FunctionSchema& schema)
 
   const auto& returns = schema.returns();
-  out << "(";
+  bool need_paren = returns.size() != 1 && !schema.is_varret();
```
What does the `is_varret()` check do? (Can you paste a before/after?)

shunting314 replied:
I actually don't know. From the name and the string representation (...), I think it means variable number of return values. Also I don't see any occurrence of `...` in `native_functions.yaml`.  The `Return.parse` method in torchgen also does not handle `...`, so I think `FunctionSchema::is_varret` is something specific to the C++ FunctionSchema. 

Not sure what does pasting 'a before/after' means here since I don't change `FunctionSchema::is_varret` itself. But now the behavior is, if there is only a single return value and `FunctionSchema.is_varret` flag is *not* on, we skip the parenthesis around the return values. This will be consistent with torchgen function schema parser.

shunting314 replied:
I only find one example of using `...` in return values here for prim ops: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/register_prim_ops.cpp#L192  .

bdhirsh replied:
oh thanks, that seems reasonable (not putting parens around `...`, and also the fact that none of our native_functions.yaml ops care about it)

bdhirsh reviewed (approved):
lgtm!

shunting314 commented:
CI failure are not related but I'll rebase

shunting314 committed:
Update on "Make string serialization of C++ FunctionSchema consistent with torchgen.model.FunctionSchema"


There is a discrepency between the string representation of C++ FunctionSchema and torchgen.model.FunctionSchema.
The latter will not add parenthesis around the returned types if that a single item,
but the C++ FunctionSchema always add the parenthesis.

Make them consistent so we can convert one type to the other via its string representation and parse method.

Differential Revision: [D36535924](https://our.internmc.facebook.com/intern/diff/D36535924/)

[ghstack-poisoned]

shunting314 committed:
Update on "Make string serialization of C++ FunctionSchema consistent with torchgen.model.FunctionSchema"


There is a discrepency between the string representation of C++ FunctionSchema and torchgen.model.FunctionSchema.
The latter will not add parenthesis around the returned types if that a single item,
but the C++ FunctionSchema always add the parenthesis.

Make them consistent so we can convert one type to the other via its string representation and parse method.

Differential Revision: [D36535924](https://our.internmc.facebook.com/intern/diff/D36535924/)

[ghstack-poisoned]

shunting314 commented:
Fix unit test failure by updating .expect files to reflect the change.

shunting314 committed:
Update on "Make string serialization of C++ FunctionSchema consistent with torchgen.model.FunctionSchema"


There is a discrepency between the string representation of C++ FunctionSchema and torchgen.model.FunctionSchema.
The latter will not add parenthesis around the returned types if that a single item,
but the C++ FunctionSchema always add the parenthesis.

Make them consistent so we can convert one type to the other via its string representation and parse method.

Differential Revision: [D36535924](https://our.internmc.facebook.com/intern/diff/D36535924/)

[ghstack-poisoned]

shunting314 committed:
Update on "Make string serialization of C++ FunctionSchema consistent with torchgen.model.FunctionSchema"


There is a discrepency between the string representation of C++ FunctionSchema and torchgen.model.FunctionSchema.
The latter will not add parenthesis around the returned types if that a single item,
but the C++ FunctionSchema always add the parenthesis.

Make them consistent so we can convert one type to the other via its string representation and parse method.

Differential Revision: [D36535924](https://our.internmc.facebook.com/intern/diff/D36535924/)

[ghstack-poisoned]

shunting314 committed:
Update on "Make string serialization of C++ FunctionSchema consistent with torchgen.model.FunctionSchema"


There is a discrepency between the string representation of C++ FunctionSchema and torchgen.model.FunctionSchema.
The latter will not add parenthesis around the returned types if that a single item,
but the C++ FunctionSchema always add the parenthesis.

Make them consistent so we can convert one type to the other via its string representation and parse method.

Differential Revision: [D36535924](https://our.internmc.facebook.com/intern/diff/D36535924/)

[ghstack-poisoned]

shunting314 committed:
Update on "Make string serialization of C++ FunctionSchema consistent with torchgen.model.FunctionSchema"


There is a discrepency between the string representation of C++ FunctionSchema and torchgen.model.FunctionSchema.
The latter will not add parenthesis around the returned types if that a single item,
but the C++ FunctionSchema always add the parenthesis.

Make them consistent so we can convert one type to the other via its string representation and parse method.

Differential Revision: [D36535924](https://our.internmc.facebook.com/intern/diff/D36535924/)

[ghstack-poisoned]

shunting314 commented:
@bdhirsh  I changed the code a bit to make existing unit tests happy.  Do you want to talk another look?

Now we skip parenthesis around the return type if
1. the return type is a single item and not varrret: e.g. -> Tensor
2. or the return type is nothing but varret. e.g. -> ...

According to the above rule some thing like `-> (Tensor, ...)` will still be enclosed with a pair of parenthesis.

There is one exception. If the return type is a single item but starts with left parenthesis, we still need add the extra prenthesis.

There are 2 cases
 1. something like 'aten::items.str(Dict(str, t) self) -> ((str, t)[])'. Without the extra parenthesis, the c++ schema parser will mistakenly treat '(str,t)[]' as returning a tuple since the lexer at most look ahead one token. That cause schema parsing fail.
 3. something like '-> ((str, str))'. Need extra parenthesis so the return type is a single tuple rather than two strings. PR (https://github.com/pytorch/pytorch/pull/23204) has more context about this. test_serialize_and_deserialize (https://github.com/pytorch/pytorch/blob/master/test/test_function_schema.py#L15) also covers this case.

shunting314 commented:
@pytorchbot land this please

shunting314 commented:
@pytorchbot merge this please

shunting314 commented:
cc @eellison for the change in case it affects TorchScript.
